src/ir_passes/recomb.ml

```diff
@@ -0,0 +1,139 @@
+  | SwitchE (exp1, cases) ->
+    SwitchE (t_exp env exp1, List.map (case env exp1) cases)
+  | TryE (exp1, cases) ->
+    TryE (t_exp env exp1, List.map (case env exp1) cases)
+  | LoopE exp1 ->
+    LoopE (t_exp env exp1)
+  | LabelE (id, typ, exp1) ->
+    LabelE (id, typ, t_exp env exp1)
+  | AsyncE (s, tb, e, typ) -> AsyncE (s, tb, t_exp env e, typ)
+  | DeclareE (id, typ, exp1) ->
+    DeclareE (id, typ, t_exp env exp1)
+  | DefineE (id, mut ,exp1) ->
+    DefineE (id, mut, t_exp env exp1)
+  | NewObjE (sort, ids, t) ->
+    NewObjE (sort, ids, t)
+  | SelfCallE (ts, e1, e2, e3) ->
+    SelfCallE (ts, t_exp env e1, t_exp env e2, t_exp env e3)
+  | ActorE (ds, fields, {meta; preupgrade; postupgrade; heartbeat; timer; inspect}, typ) ->
+    (* Until Actor expressions become their own units,
+       we repeat what we do in `comp_unit` below *)
+    let env1 = empty_env () in
+    let ds' = t_decs env1 ds in
+    let preupgrade' = t_exp env1 preupgrade in
+    let postupgrade' = t_exp env1 postupgrade in
+    let heartbeat' = t_exp env1 heartbeat in
+    let timer' = t_exp env1 timer in
+    let inspect' = t_exp env1 inspect in
+    ActorE (ds', fields,
+      {meta;
+       preupgrade = preupgrade';
+       postupgrade = postupgrade';
+       heartbeat = heartbeat';
+       timer = timer';
+       inspect = inspect'
+      }, typ)
+
+and t_lexp env (e : Ir.lexp) = { e with it = t_lexp' env e.it }
+and t_lexp' env = function
+  | VarLE id -> VarLE id
+  | IdxLE (exp1, exp2) ->
+    IdxLE (t_exp env exp1, t_exp env exp2)
+  | DotLE (exp1, n) ->
+    DotLE (t_exp env exp1, n)
+
+and t_dec env dec = { dec with it = t_dec' env dec.it }
+
+and t_dec' env dec' =
+  match dec' with
+  | LetD (pat,exp) -> LetD (pat,t_exp env exp)
+  | VarD (id, typ, exp) -> VarD (id, typ, t_exp env exp)
+  | RefD (id, typ, lexp) -> RefD (id, typ, t_lexp env lexp)
+
+and t_decs env decs = List.map (t_dec env) decs
+
+and t_block env (ds, exp) = (t_decs env ds, t_exp env exp)
+
+and t_comp_unit = function
+  | LibU _ -> raise (Invalid_argument "cannot compile library")
+  | ProgU ds ->
+    let env = empty_env () in
+    let ds' = t_decs env ds in
+    ProgU  ds'
+  | ActorU (as_opt, ds, fields, {meta; preupgrade; postupgrade; heartbeat; timer; inspect}, typ) ->
+    let env = empty_env () in
+    let ds' = t_decs env ds in
+    let preupgrade' = t_exp env preupgrade in
+    let postupgrade' = t_exp env postupgrade in
+    let heartbeat' = t_exp env heartbeat in
+    let timer' = t_exp env timer in
+    let inspect' = t_exp env inspect in
+    ActorU (as_opt, ds', fields,
+      {meta;
+       preupgrade = preupgrade';
+       postupgrade = postupgrade';
+       heartbeat = heartbeat';
+       timer = timer';
+       inspect = inspect'
+      }, typ)
+
+(* Entry point for the program transformation *)
+
+let transform (cu, flavor) =
+  assert (not flavor.has_typ_field); (* required for hash_typ *)
+  (t_comp_unit cu, {flavor with has_poly_eq = false})
```

Comment:
```suggestion
  (t_comp_unit cu, {flavor with has_poly_eq = false})
```
Copy paster remnant? (has_poly_eq)

Reply:
not sure, `Ir_def.prog` is `comp_unit * flavor`, I guess we don't chang the flavour, so
``` OCaml
let transform (cu, flavor) =
  (t_comp_unit cu, flavor)
```
?

Reply:
Fixed: 863946604